### PR TITLE
feat: Migrate LinkReferenceProvider to OCP

### PR DESCRIPTION
**Pull request by kyteinsky**
## Summary
This would be useful to eleminate the need for using this OC class
when an app dev wants to implement a custom reference provider
for the web client but wants to fall back on opengraph for mobile
clients.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by kyteinsky** [comment]
It could've been an abstract class in OCP but that would interfere with `ADiscoverableReferenceProvider.php` so that idea was dropped.  
Please squash before merge. The two commits are separated for ease of review only.

File: lib/private/Collaboration/Reference/LinkReferenceProvider.php [hunk 1]
```diff
@@ -24,160 +25,8 @@
...
-	}
-
-	public function getCachePrefix(string $referenceId): string {
-		return $referenceId;
-	}
-
-	public function getCacheKey(string $referenceId): ?string {
-		return null;
-	}
+/** @deprecated 29.0.0 */
```
Comment (julien-nc) [inline lib/private/Collaboration/Reference/LinkReferenceProvider.php:30]:
```suggestion
/** @deprecated 29.0.0 Use OCP\Collaboration\Reference\LinkReferenceProvider instead */
```

**Review by julien-nc — CHANGES_REQUESTED** [review]
Minor comment.

Then :+1: :blue_heart:

**Review by julien-nc — APPROVED** [review]

**Review by juliusknorr — APPROVED** [review]

**Comment by juliushaertl** [comment]
> The autoloaders are not up to date
> Please run: bash build/autoloaderchecker.sh

**Comment by kyteinsky** [comment]
there are a lot of changed hashes in the apps. Should that be included in the commit?  
```diff
apps/admin_audit/composer/composer/installed.php

-        'reference' => 'b1797842784b250fb01ed5e3bf130705eb94751b',
+        'reference' => '49969705de01aff43cf2e6b662e5c96bd48a3b27',
```

**Pull request merged**